rfcs/20190815-tfx-notebook.md

```diff
@@ -0,0 +1,563 @@
+
+##### Part 2 (Option 2): Artifact execution via show(artifact)
+
+In this alternative, instead of running components and retrieving artifacts
+after they are run, artifacts are "run" implicitly when show(artifact) is
+called. This will implicitly execute the component necessary to generate the
+artifact.
+
+Pros: One show() call rather than separate run() and show(). Dependencies can be
+handled under the hood, and we can avoid visualizing stale results.
+
+Cons: Not intuitive as this is not what “show” means. Not the simplest mental
+model and potentially confusing. If a user wants to always show artifacts after
+running, it is very natural to put run() and show() in the same notebook cell.
+Running code and components that are not part of the current executed cell is
+also not a notebook-friendly pattern.
+
+##### Part 3: Notebook visualizations for well-known artifact types can be registered
+
+We introduce a `NotebookArtifactVisualizationRegistry` class on which we may
+register visualizations (e.g. HTML renderings for Colab / Jupyter notebooks),
+which are to be returned from ExecutionResult.read() when run in the notebook
+environment. For specific artifact types, we allow registration of handlers to
+return visualizations for those types. We will write visualizations for
+well-known artifact types we use. For example, the `ExampleStatistics` Artifact
+type output by StatisticsGen could be visualized by producing an interactive
+display of the resulting statistics
+[using Facets](https://pair-code.github.io/facets/).
+
+##### Example notebook usage
+
+Here is an example of what notebook execution may look like in this scheme.
+
+```
+Input[0]:
+# To begin, we initialize an interactive context. Here, by not passing
+# in a base directory or metadata configuration, we create an ephemeral
+# context whose outputs will be in a temporary directory.
+context = InteractiveContext()
+
+# Alternatively, we may pass in these directories for a context using a
+# persistent store:
+#
+# context = InteractiveContext(base_dir=my_base_dir,
+#                              metadata_connection_config=my_config)
+```
+
+<table>
+  <tr>
+   <td><strong>Input[1]:</strong>
+<p>
+<code><em># First, ExampleGen with a run / read.</em></code>
+<code>example_gen = CsvExampleGen(input_base=examples)</code>
+<p>
+<code><em># Note that the output will be of type 'ExamplesPath', for which we</em></code>
+<code><em># may have registered a notebook visualization handler.</em></code>
+<code>example_gen_result = context.run(example_gen)</code>
+<p>
+<code>example_gen.outputs['examples']</code>
+<p>
+<code><em># alternative style: explicit context.show() method</em></code>
+<code>context.show(example_gen.outputs['examples'])</code>
+   </td>
+  </tr>
+  <tr>
+   <td><strong>Output[1]:</strong>
+<p>
+_notebook visualization indicating we have N examples at some temp path_
+   </td>
+  </tr>
+</table>
```

Comment:
Nit: I don't know what happened to the styling here, but it makes this section very hard to read. Consider making all of this a code block and separating the "cells" with comments or something?